src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+                {
+                    return true;
+                }
+
+                if (thisEnumerator.Current != otherEnumerator.Current)
+                {
+                    return false;
+                }
+            }
+        }
+
+        public static bool operator ==(Utf8Span left, Utf8Span right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Utf8Span left, Utf8Span right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static bool operator ==(Utf8Span left, string right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Utf8Span left, string right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static bool operator ==(string left, Utf8Span right)
+        {
+            return right.Equals(left);
+        }
+
+        public static bool operator !=(string left, Utf8Span right)
+        {
+            return !right.Equals(left);
+        }
+
+        public int CompareTo(Utf8Span other)
+        {
+            throw new NotImplementedException();
+        }
+
+        public int CompareTo(string other)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="index">Index in UTF-8 code units (bytes)</param>
+        /// <returns>Length in UTF-8 code units (bytes)</returns>
+        public Utf8Span Substring(int index)
+        {
+            return Substring(index, Length - index);
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="index">Index in UTF-8 code units (bytes)</param>
+        /// <returns>Length in UTF-8 code units (bytes)</returns>
+        public Utf8Span Substring(int index, int length)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException("index");
+            }
+
+            if (length < 0)
+            {
+                // TODO: Should we support that?
+                throw new ArgumentOutOfRangeException("length");
+            }
+
+            if (length == 0)
+            {
+                return Empty;
+            }
+
+            if (length == Length)
+            {
+                return this;
+            }
+
+            if (index + length > Length)
+            {
+                // TODO: Should this be index or length?
+                throw new ArgumentOutOfRangeException("index");
+            }
+
+            return new Utf8Span(_buffer.Slice(index, length));
+        }
+
+        // TODO: Naive algorithm, reimplement faster
+        // TODO: Should this be public?
+        public int IndexOf(Utf8Span value)
+        {
+            if (value.Length == 0)
+            {
+                // TODO: Is this the right answer?
+                // TODO: Does this even make sense?
+                return 0;
```

Comment:
This is correct if we want to match the behavior with string index of:
https://msdn.microsoft.com/en-us/library/k8b1470s(v=vs.110).aspx
> The zero-based index position of value if that string is found, or -1 if it is not. If value is String.Empty, the return value is 0.